### ISSUE TITLE
Possible invalid type error with overloaded AnyStr

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#3817]
The following code gives a type error:

```python
import os
from typing import AnyStr, overload

@overload
def splitext_all(p: os.PathLike[AnyStr]) -> tuple[AnyStr, AnyStr]: ...
@overload
def splitext_all(p: AnyStr) -> tuple[AnyStr, AnyStr]: ...

def splitext_all(p):
    f = os.fspath(p)

    return f, f
```

At `def splitext_app(p):`

![image](https://user-images.githubusercontent.com/376842/184128363-fe63d997-7eca-4dea-b8ee-e3d869b9a31b.png)

As far as I can tell it should be correct, and the types for `os.fspath()` and `os.PathLike` look right to me. First time I've used `@overload` though so sorry if I'm doing something stupid.

It also infers the type of `f` as `str` which doesn't seem right:

![image](https://user-images.githubusercontent.com/376842/184128576-452b63fe-ac0e-47ba-9b27-4e3be527902b.png)

This is with v2022.12.0 of the Python extension in VSCode.

By the way what does the `*` mean in `bytes*`?

**Comment by erictraut** [comment]
Pyright is working as designed here. This isn't a bug.

First, `AnyStr` is a TypeVar, and you need to include it in the signature of the overload implementation.

```python
def splitext_all(p: AnyStr | os.PathLike[AnyStr]) -> tuple[AnyStr, AnyStr]:
    ...
```

The `*` indicates that it's a [conditional type](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#constrained-type-variables-and-conditional-types).

**Comment by Timmmm** [comment]
Ah yeah actually there's no need for the overload in this case. But this doesn't work either:

```
import os
from typing import AnyStr

def splitext_all(p: AnyStr | os.PathLike[AnyStr]) -> tuple[AnyStr, AnyStr]:
    f = os.fspath(p)

    return f, f
```

<img width="723" alt="image" src="https://user-images.githubusercontent.com/376842/184190007-9a6d5226-4313-490e-a22a-b0bc04e5f474.png">

`fspath` is

```
@overload
def fspath(path: str) -> str: ...
@overload
def fspath(path: bytes) -> bytes: ...
@overload
def fspath(path: PathLike[AnyStr]) -> AnyStr: ...
```

(Which I guess could also just be `def fspath(path: AnyStr | PathLike[AnyStr]) -> AnyStr: ...`?)

Also [the example in the official docs](https://docs.python.org/3/library/typing.html#typing.overload) doesn't repeat the type annotations for the actual implementation - is that wrong?

Thanks for the help!

**Comment by bschnurr** [comment]
this sort of works
```python
def splitext_all(p: AnyStr | os.PathLike[AnyStr]) -> tuple[AnyStr, AnyStr]:
    if isinstance(p, os.PathLike):
         f = os.fspath(p)
         return f , f
    f   = os.fspath(p)

    return f, f
```

**Comment by Timmmm** [comment]
Can you actually use `isinstance` with `PathLike` though? I mean I guess it doesn't technically matter in that case since both branches are the same... but still. Doesn't seem right.

**Comment by erictraut** [comment]
Here's a simplified example that demonstrates what's happening:

```python
from typing import TypeVar, overload

T = TypeVar("T", str, bytes)

@overload
def func1(x: str) -> str: ...
@overload
def func1(x: bytes) -> bytes: ...
def func1(x: str | bytes) -> str | bytes:
    ...

def func2(y: T):
    func1(y)
```

Pyright performs "union expansion" as part of overload matching. It doesn't currently take into account constrained type variables when performing this union expansion because (technically) constrained TypeVars are not unions, but they do act somewhat like them.

I've extended support for union expansion during overload matching to support this use case. This will be included in the next release.

**Comment by Timmmm** [comment]
Thanks! Don't burn out!

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.267, which I just published. It will also be included in a future version of pylance.
